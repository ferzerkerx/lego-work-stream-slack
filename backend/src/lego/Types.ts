import { LegoSelectMessage } from './LegoSelectMessage';
import { Metrics } from './metrics/Metrics';
import { SlackMessage } from 'botkit';

export interface LegoSelectMessageRepository {
  findMessagesBy(config: MetricsConfiguration): Promise<LegoSelectMessage[]>;

  find(messageId: string): Promise<LegoSelectMessage>;

  save(legoMessage: LegoSelectMessage): Promise<LegoSelectMessage>;
}

export interface LegoMetricsService {
  metricsForConfig(config: MetricsConfiguration): Promise<Metrics>;
}

export interface LegoSelectionReplyService {
  createReply(message): Promise<SlackMessage | void>;
}
