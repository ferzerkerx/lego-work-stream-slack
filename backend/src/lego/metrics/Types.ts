export interface CategoryValueMap {
  [category: string]: number;
}
