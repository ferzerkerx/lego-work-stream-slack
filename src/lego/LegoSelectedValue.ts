import { LegoSelectedValueEntry } from './LegoSelectedValueEntry';

export class LegoSelectedValue {
  id: string;
  entries: LegoSelectedValueEntry[] = [];
}
