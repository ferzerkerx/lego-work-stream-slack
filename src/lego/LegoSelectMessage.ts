import {LegoSelectedValue} from "./LegoSelectedValue";

export class LegoSelectMessage {
  id: string;
  selectedValues: LegoSelectedValue[] = [];
  channel: string;
  date: Date = new Date();
}