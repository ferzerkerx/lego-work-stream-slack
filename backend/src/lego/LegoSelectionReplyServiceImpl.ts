import { SlackAttachment, SlackMessage } from 'botkit';
import { LegoSelectMessage } from './LegoSelectMessage';
import { LegoSelectionService } from './LegoSelectionService';
import { LegoSelectedValue } from './LegoSelectedValue';
import { ErrorUtil } from '../utils/ErrorUtil';
import {
  LegoSelectionReplyService,
  LegoSelectMessageRepository,
} from './Types';

export class LegoSelectionReplyServiceImpl
  implements LegoSelectionReplyService {
  constructor(
    private legoSelectMessageRepository: LegoSelectMessageRepository
  ) {}

  createReply(message): Promise<SlackMessage | void> {
    let fullMessageId: string = LegoSelectionReplyServiceImpl.getMessageId(
      message
    );

    return this.legoSelectMessageRepository
      .find(fullMessageId)
      .then((storedLegoMessage: LegoSelectMessage) => {
        const legoMessage = LegoSelectionService.createLegoSelectMessage({
          legoMessage: storedLegoMessage,
          messageContext: {
            fullMessageId,
            userData: {
              userId: message.user,
              userName: message.raw_message.user.name,
            },
            action: message.actions[0],
            channelData: {
              channelId: message.channel,
              name: message.raw_message.channel.name,
            },
          },
        });

        return this.legoSelectMessageRepository.save(legoMessage).then(() => {
          return LegoSelectionReplyServiceImpl._createReply(
            message,
            legoMessage
          );
        });
      })
      .catch(e => ErrorUtil.defaultErrorHandling(e));
  }

  private static getMessageId(message): string {
    return `${message.channel}_${message.original_message.ts}`;
  }

  private static _createReply(
    message,
    legoMessage: LegoSelectMessage
  ): SlackMessage {
    const reply: SlackMessage = message.original_message;

    let attachmentsToSend: SlackAttachment[] = reply.attachments.filter(
      attachment => attachment.callback_id === 'lego_stats'
    );

    let actionMappings: any = {};
    attachmentsToSend[0].actions.forEach(actionDetail => {
      actionMappings[actionDetail.name] = { text: actionDetail.text };
    });

    attachmentsToSend.push({
      text: `${this.formatSelectedValues(
        legoMessage.selectedValues,
        actionMappings
      )}`,
    });

    reply.attachments = attachmentsToSend;

    return reply;
  }

  static formatSelectedValues(
    legoSelectedValues: LegoSelectedValue[],
    actionMappings: any
  ): string {
    let messageStr: string = '';

    for (let legoSelectedValue of legoSelectedValues) {
      let entriesStr: string = legoSelectedValue.entries
        .filter(entry => entry.value > 0)
        .map(entry => `<@${entry.userData.userId}>:${entry.value}`)
        .join(',');

      const actionDisplayName = actionMappings[legoSelectedValue.id]
        ? actionMappings[legoSelectedValue.id].text
        : legoSelectedValue.id;
      if (entriesStr.length > 0) {
        messageStr += `${actionDisplayName}: ${entriesStr}\n`;
      }
    }
    return messageStr;
  }
}
