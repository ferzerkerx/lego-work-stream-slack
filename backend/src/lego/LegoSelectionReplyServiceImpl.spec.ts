import { LegoSelectedValue } from './LegoSelectedValue';
import { LegoSelectionReplyServiceImpl } from './LegoSelectionReplyServiceImpl';
import { LegoSelectMessageRepository } from './Types';
import { LegoSelectMessage } from './LegoSelectMessage';

describe('LegoSelectionReplyService', () => {
  test('formatMessage properly', () => {
    let selectedValues: LegoSelectedValue[] = [
      {
        id: 'someActionName',
        entries: [
          {
            userData: { userId: 'someuserId', userName: 'someuser' },
            value: 8,
          },
        ],
      },
      {
        id: 'someOtherActionName',
        entries: [
          {
            userData: { userId: 'someuserId', userName: 'someuser' },
            value: 6,
          },
        ],
      },
      {
        id: 'awesomeActionName',
        entries: [
          {
            userData: { userId: 'someuserId', userName: 'someuser' },
            value: 0,
          },
        ],
      },
    ];

    let actionMappings: any = {
      someActionName: { text: 'theActualName' },
      awesomeActionName: { text: 'CoolName' },
    };

    const legoSelectMessage: string = LegoSelectionReplyServiceImpl.formatSelectedValues(
      selectedValues,
      actionMappings
    );
    expect(legoSelectMessage).toMatchSnapshot();
  });

  test('createReply', () => {
    const legoSelectionReplyServiceImpl = new LegoSelectionReplyServiceImpl(
      new FakeRepo()
    );
    const message = {
      ts: 12345,
      channel: 'channelName',
      original_message: {
        ts: 12345,
        attachments: [
          {
            callback_id: 'lego_stats',
            actions: [{ name: 'someActionName', text: 'someActionText' }],
          },
        ],
      },
      user: 'someUserId',
      raw_message: {
        user: { name: 'someuser' },
        channel: { name: 'channelName' },
      },
      actions: [{ name: 'someActionName', selected_options: [{ value: 8 }] }],
    };

    legoSelectionReplyServiceImpl.createReply(message).then(slackMessage => {
      expect(slackMessage).toMatchSnapshot();
    });
  });

  class FakeRepo implements LegoSelectMessageRepository {
    find(messageId: string): Promise<LegoSelectMessage> {
      const legoSelectMessage = new LegoSelectMessage();
      legoSelectMessage.selectedValues.push({
        id: 'someActionName',
        entries: [
          {
            userData: { userId: 'someUserId', userName: 'someuser' },
            value: 5,
          },
        ],
      });

      return Promise.resolve(legoSelectMessage);
    }

    findMessagesBy(config: MetricsConfiguration): Promise<LegoSelectMessage[]> {
      return Promise.resolve([]);
    }

    save(legoMessage: LegoSelectMessage): Promise<LegoSelectMessage> {
      return Promise.resolve(legoMessage);
    }
  }
});
