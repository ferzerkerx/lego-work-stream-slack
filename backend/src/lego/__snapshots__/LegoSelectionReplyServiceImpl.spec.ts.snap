// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LegoSelectionReplyService createReply 1`] = `
Object {
  "attachments": Array [
    Object {
      "actions": Array [
        Object {
          "name": "someActionName",
          "text": "someActionText",
        },
      ],
      "callback_id": "lego_stats",
    },
    Object {
      "text": "someActionText: <@someUserId>:8
",
    },
  ],
  "ts": 12345,
}
`;

exports[`LegoSelectionReplyService formatMessage properly 1`] = `
"theActualName: <@someuserId>:8
someOtherActionName: <@someuserId>:6
"
`;
